larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc

```diff
@@ -0,0 +1,500 @@
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "canvas/Utilities/InputTag.h"
+#include "canvas/Persistency/Common/FindManyP.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+// LArSoft libraries
+#include "nusimdata/SimulationBase/MCParticle.h"
+#include "nusimdata/SimulationBase/MCTruth.h"
+#include "larcoreobj/SimpleTypesAndConstants/RawTypes.h" // raw::ChannelID_t
+#include "larcore/Geometry/Geometry.h"
+#include "larcorealg/Geometry/PlaneGeo.h"
+#include "lardataobj/Simulation/SimChannel.h"
+#include "lardataobj/RawData/RawDigit.h"
+#include "lardataobj/RawData/raw.h"
+#include "lardataobj/RecoBase/Hit.h"
+#include "larsim/MCCheater/ParticleInventoryService.h"
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+
+// DUNETPC specific includes
+//#include "dune/DAQTriggerSim/TriggerDataProducts/TriggerTypes.h"
+//#include "dune/DAQTriggerSim/TriggerDataProducts/BasicTrigger.h"
+//#include "dune/DuneInterface/AdcTypes.h"
+//#include "dune/DuneInterface/SimChannelExtractService.h"
+
+#include "c2numpy.h"
+
+using std::string;
+using std::cout;
+using std::endl;
+using std::ofstream;
+
+struct WireSigInfo {
+  int pdgcode;
+  std::string genlab;
+  std::string procid;
+  unsigned int tdcmin;
+  unsigned int tdcmax;
+  int numel;
+  double edep;
+};
+
+namespace nnet {
+  class RawWaveformDump;
+}
+
+class nnet::RawWaveformDump : public art::EDAnalyzer {
+
+public:
+
+  explicit RawWaveformDump(fhicl::ParameterSet const& p);
+
+  // Plugins should not be copied or assigned.
+  RawWaveformDump(RawWaveformDump const&) = delete;
+  RawWaveformDump(RawWaveformDump&&) = delete;
+  RawWaveformDump& operator=(RawWaveformDump const&) = delete;
+  RawWaveformDump& operator=(RawWaveformDump&&) = delete;
+
+  // Required functions.
+  void analyze(art::Event const& e) override;
+
+  void reconfigure(fhicl::ParameterSet const & p);
+
+  void beginJob() override;
+  void endJob() override;
+
+private:
+
+  std::string fDumpWaveformsFileName;
+
+  std::string	fSimulationProducerLabel; ///< The name of the producer that tracked simulated particles through the detector
+  std::string	fDigitModuleLabel; ///< module that made digits
+
+  std::string fSelectGenLabel;
+  std::string fSelectProcID;
+  int fSelectPDGCode;
+  std::string fPlaneToDump;
+  double fMinParticleEnergyGeV;
+  double fMinEnergyDepositedMeV;
+  int fMinNumberOfElectrons;
+  int fMaxNumberOfElectrons;
+  bool fSaveSignal;
+  art::ServiceHandle<geo::Geometry> fgeom;
+  art::ServiceHandle<cheat::ParticleInventoryService> PIS;
+  //art::ServiceHandle<SimChannelExtractService> m_pscx;
+  detinfo::DetectorClocks const * fClks;
```

Comment:
Are any of these needed as data members or could they be created on the fly as they are all services?

Reply:
No, it is not needed but there is nothing wrong with this implementation.